### PR TITLE
[WIP] - tfe.keras - logistic regression

### PR DESCRIPTION
**Pull request by yanndupis**
Hey,

#540 
This is a WIP PR to support simple models (e.g. logistic regression) with tfe.keras interface. 

There are several remaining items:
- [ ] `.fit()` expects private input right now but it should work with numpy array as well. In a separate PR we could support `tf.data.Dataset` by creating a TFE version of it.
- [ ] support accuracy
- [ ] potentially use callbacks to modularize `fit` step and track loss, accuracy etc. https://keras.io/callbacks/
- [ ] add doctrings to optimizers, loss etc. 
- [ ] in binary cross entropy support `from_logit` parameter
- [ ] refactor gradient tests for Dense and sigmoid layer to make it cleaner.

### CODE REVIEW

File: tf_encrypted/keras/losses.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        binary_crossentropy)
+
+  def grad(self, y_true, y_pred):
+    return y_pred - y_true
+
+def binary_crossentropy(y_true, y_pred):
+
+  batch_size = y_true.shape.as_list()[0]
+  batch_size_inv = 1 / batch_size
+  out = y_true * get_protocol().log(y_pred)
```
Comment (mortendahl) [inline tf_encrypted/keras/losses.py]: maybe we can't do it now but eventually (ie after refactoring) we should get rid of the `get_protocol` call and simply use `tfe.log(..)` instead (which will find the right protocol)

File: tf_encrypted/keras/losses.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  out += (y_true.negative() + 1) * get_protocol().log(y_pred.negative() + 1)
```
Comment (mortendahl) [inline tf_encrypted/keras/losses.py]: does `1 - y_true` work?

File: examples/notebooks/keras-training/common.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
+"""Provide classes to perform private training and private prediction with
+logistic regression"""
+import tensorflow as tf
+# pylint:  disable=redefined-outer-name
+import tf_encrypted as tfe
+
+
+class LogisticRegression:
```
Comment (mortendahl) [inline examples/notebooks/keras-training/common.py]: maybe it's time we move this guy into its own directory, say `tfe.keras.applications` ([to match](https://www.tensorflow.org/api_docs/python/tf/keras/applications))

File: examples/notebooks/keras-training/data.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+"""Dummy data preparation for logistic regression training."""
```
Comment (mortendahl) [inline examples/notebooks/keras-training/data.py]: is this file used anywhere?

File: tf_encrypted/keras/models/sequential.py [hunk 5]
```diff
@@ -106,6 +110,33 @@ def layers(self):
...
+      grad_weights, d_y = layer.backward(d_y)
+      self._optimizer.apply_gradients(layer.weights, grad_weights)
+
+  def compile(self, optimizer, loss=None):
+    self._optimizer = optimizers.get(optimizer)
+    self._loss = loss
+
+  def fit_batch(self, x, y):
+    y_pred = self.call(x)
+    dy = self._loss.grad(y, y_pred)
```
Comment (mortendahl) [inline tf_encrypted/keras/models/sequential.py]: what happens if you compiled with `loss=None`?

File: tf_encrypted/keras/models/sequential.py [hunk 6]
```diff
@@ -106,6 +110,33 @@ def layers(self):
...
+    self.backward(dy)
+    loss = self._loss(y, y_pred)
+    sess = KE.get_session()
+    self._current_loss = sess.run(loss.reveal())
+
+  def fit(self, x, y, epochs=1, steps_per_epoch=1):
+    for e in range(epochs):
+      print('Epoch {}/{}'.format(e + 1, epochs))
```
Comment (mortendahl) [inline tf_encrypted/keras/models/sequential.py]: use logger instead (if anything)

File: tf_encrypted/keras/optimizers.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+from tensorflow.keras import backend as K
+
+import tf_encrypted as tfe
+
+class SGD():
+  def __init__(self, lr=0.01):
+    self.lr = lr
+
+  def apply_gradients(self, var, grad):
+    sess = K.get_session()
```
Comment (mortendahl) [inline tf_encrypted/keras/optimizers.py]: wouldn't this give you a `tf.Session` instead of `tfe.Session`?

File: examples/notebooks/keras-training/data.py (hunk 4 above)
Comment (yanndupis) [reply examples/notebooks/keras-training/data.py]: Fixed - deleted, thanks

File: tf_encrypted/keras/optimizers.py (hunk 7 above)
Comment (yanndupis) [reply tf_encrypted/keras/optimizers.py]: Fixed - good catch

File: tf_encrypted/keras/models/sequential.py (hunk 5 above)
Comment (yanndupis) [reply tf_encrypted/keras/models/sequential.py]: I think it will crash :)

File: examples/notebooks/keras-training/common.py (hunk 3 above)
Comment (yanndupis) [reply examples/notebooks/keras-training/common.py]: Not sure. I think this repo is focusing on pre-trained models: https://keras.io/applications/

File: tf_encrypted/keras/models/sequential.py (hunk 6 above)
Comment (yanndupis) [reply tf_encrypted/keras/models/sequential.py]: Fixed -  I used logger instead. But it looks quite ugly lol. Maybe there is the option to remove the pink color at least

File: tf_encrypted/keras/backend.py [hunk 8]
```diff
@@ -20,9 +20,9 @@ def get_session(op_input_list=()):
   default_session = ops.get_default_session()
   if default_session is not None:
     # If the default session is a TFE Session return this session
-    if isinstance(default_session, tfe.Session()):
+    if isinstance(default_session, tfe.Session):
       return default_session
-    if not isinstance(default_session, tfe.Session()):
+    if not isinstance(default_session, tfe.Session):
```
[the remark this answers is not in the document]
Comment (yanndupis) [reply tf_encrypted/keras/backend.py]: Fixed

File: tf_encrypted/keras/__init__.py [hunk 9]
```diff
@@ -14,4 +15,5 @@
     'layers',
     'models',
     'Sequential',
+    'losses'
```
[the remark this answers is not in the document]
Comment (yanndupis) [reply tf_encrypted/keras/__init__.py]: Fixed

File: tf_encrypted/keras/layers/activation.py [hunk 10]
```diff
@@ -18,12 +18,20 @@ class Activation(Layer):
...
+    y = self.activation(inputs)
+    self._layer_output = y
+    return y
 
   def compute_output_shape(self, input_shape):
     return input_shape
+
+  def backward(self, d_y):
+    grad_weights = []
+    d_x = d_y * self.activation_deriv(self._layer_output)
```
[the remark this answers is not in the document]
Comment (yanndupis) [reply tf_encrypted/keras/layers/activation.py]: Fixed

File: tf_encrypted/keras/losses.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
+"""TFE Keras loss function"""
+from tf_encrypted import get_protocol
+
+class Loss():
+  """Loss base class."""
+  def __init__(self,
+               fn,
+               **kwargs):
+
+    self.fn = fn
```
[the remark this answers is not in the document]
Comment (yanndupis) [reply tf_encrypted/keras/losses.py]: Fixed

File: tf_encrypted/keras/losses.py (hunk 1 above)
Comment (yanndupis) [reply tf_encrypted/keras/losses.py]: Agreed

File: tf_encrypted/keras/losses.py (hunk 2 above)
Comment (yanndupis) [reply tf_encrypted/keras/losses.py]: It does, nice! I didn't know 1 will be lifted this way as well when the other tensor is a private tensor.

**Comment by yanndupis** [comment]
Thank you @mortendahl  for your review! I have addressed most of the comments (noted Fixed). Not sure I will have much time to work on this next week. So feel free to push some code if needed to get it merge. Thanks again.

**Comment by mortendahl** [comment]
Alright, I can take over for now, thanks @yanndupis!

File: examples/notebooks/keras-training/common.py (hunk 3 above)
Comment (mortendahl) [reply examples/notebooks/keras-training/common.py]: makes sense 👍

**Comment by mortendahl** [comment]
A larger issue is how to handle gradients. The current code allows eg activation functions to be given, but doesn't handle derivatives correctly. Not sure what the best approach is here, maybe follow the pattern of `tf.custom_gradient` where two functions are returned?

**Comment by yanndupis** [comment]
Hey @mortendahl , what do you mean by "eg activation functions to be given, but doesn't handle derivatives correctly.". 

one thing that the code is not handling is if the activation is specified in the Dense layer (e.g. tfe.keras.layers.Dense(1, activation='sigmoid'), it will work only if you do `tfe.keras.layers.Dense(1) then `tfe.keras.layers.Activation('sigmoid')

Are you referring to something else?

**Comment by yanndupis** [comment]
@mortendahl, we can use this issue: https://github.com/tf-encrypted/tf-encrypted/issues/540
